src/host-config/index.ts

```diff
@@ -57,75 +62,149 @@ export function parse(
+		(typeof targetConf.redsocks === 'object' &&
+			Object.keys(targetConf.redsocks).length === 0)
+	) {
+		delete patchedConf.redsocks;
+		delete patchedConf.dns;
+	}
+
 	return patchedConf;
 }
 
 export async function patch(
-	conf: HostConfiguration | LegacyHostConfiguration,
+	targetConf: HostConfiguration | LegacyHostConfiguration,
 	force: boolean = false,
 ): Promise<void> {
 	const apps = await applicationManager.getCurrentApps();
 	const appIds = Object.keys(apps).map((strId) => parseInt(strId, 10));
 
-	if (conf.network.hostname != null) {
-		await setHostname(conf.network.hostname);
+	const { network: target } = targetConf;
+
+	if (target.hostname != null) {
+		await setHostname(target.hostname);
 	}
 
-	if (conf.network.proxy != null) {
-		const targetConf = conf.network.proxy;
+	if (target.proxy != null || target.dns != null) {
 		// It's possible for appIds to be an empty array, but patch shouldn't fail
 		// as it's not dependent on there being any running user applications.
 		return updateLock.lock(appIds, { force }, async () => {
-			const proxyConf = await readProxy();
-			let currentConf: ProxyConfig | undefined = undefined;
-			if (proxyConf) {
-				delete proxyConf.noProxy;
-				currentConf = proxyConf;
+			// Get current proxy config
+			const current = await readProxy();
+			const currentConf: Partial<RedsocksConfig> = {};
+			if (current) {
+				const { proxy, dns } = current;
+				if (proxy) {
+					delete proxy.noProxy;
+					currentConf.redsocks = proxy;
+				}
+				if (dns) {
+					currentConf.dns = dns;
+				}
+			}
+
+			const toPatch: {
+				redsocks?: Partial<ProxyConfig> | null;
+				dns?: DnsConfig | null;
+			} = {};
+
+			// Extract target proxy from input
+			if (target.proxy) {
+				// target.proxy could be Record<string, any> due to legacy support,
+				// so only extract only defined fields that are part of ProxyConfig
+				const { type, ip, port, login, password } = target.proxy;
+				const definedConf = Object.fromEntries(
+					Object.entries({ type, ip, port, login, password }).filter(
+						([, value]) => value,
+					),
+				) as ProxyConfig;
+				if (Object.keys(definedConf).length > 0) {
+					toPatch.redsocks = definedConf;
+				} else {
+					// Mark for deletion if target proxy is empty
+					toPatch.redsocks = null;
+				}
+			}
+
+			// Extract target dns from input
+			if (typeof target.dns === 'boolean') {
+				if (target.dns) {
+					// If dns is true, set to default dns config
+					toPatch.dns = {
+						remote_ip: DEFAULT_REMOTE_IP,
+						remote_port: DEFAULT_REMOTE_PORT,
+					};
+				} else {
+					// Mark for deletion if target dns is false
+					toPatch.dns = null;
+				}
+			}
+			if (typeof target.dns === 'string') {
+				// Convert dns string into DnsConfig object
+				const [ip, port] = target.dns.split(':');
+				toPatch.dns = {
+					remote_ip: ip || DEFAULT_REMOTE_IP,
+					remote_port: port ? parseInt(port, 10) : DEFAULT_REMOTE_PORT,
+				};
```

Comment:
Perhaps we could move this to a separate function within the module for readability?

Reply:
Moved to `function dnsToDnsu2t` [here](https://github.com/balena-os/balena-supervisor/pull/2284/commits/39efabcccc28605b2ba3b538d48cf37b6cee0dcd#diff-87da4178ce07f20bae291890a95e15b8bf7990d89aeb29ec6bfe1f77b508c3e5R171)